File: nodejs/tests/transport/uart_test.js
'use strict';

const { UARTTransport } = require('../../packages/periph/src/transport/uart');

const UART_PORT = process.env.UART_PORT || '/dev/ttyS0';
const UART_BAUD = parseInt(process.env.UART_BAUD || '9600', 10);

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else           { console.log('FAIL', label); failed++; }
}

// Assumes a loopback jumper bridging TXD and RXD on the UART port under test.
async function main() {
    const transport = new UARTTransport(UART_PORT, { baudRate: UART_BAUD });
    await transport.open();

    await transport.write(Buffer.from([0x42]));
    checkTrue('write accepted', true);

    const rxByte = await transport.read(1);
    checkTrue('read returns 1 byte', rxByte.length === 1);
    checkTrue('loopback byte matches', rxByte[0] === 0x42);

    const resp = await transport.writeRead(Buffer.from([0xA5, 0x5A]), 2);
    checkTrue('writeRead returns 2 bytes', resp.length === 2);
    checkTrue('writeRead loopback matches', resp[0] === 0xA5 && resp[1] === 0x5A);

    await transport.close();

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main().catch(err => { console.error(err); process.exit(1); });

File: python/tests/transport/uart_test.py
import _testconfig as cfg
from machine import UART, Pin
from periph.transport.uart_micropython import UARTTransport

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# Assumes a loopback jumper bridging TXD and RXD on the UART port.
uart = UART(cfg.UART_ID, baudrate=cfg.UART_BAUD, tx=Pin(cfg.UART_TX), rx=Pin(cfg.UART_RX),
            timeout=1000)
transport = UARTTransport(uart, cfg.UART_BAUD)

transport.write(bytes([0x42]))
check_true('write accepted', True)

data = transport.read(1)
check_true('read returns 1 byte', len(data) == 1)
check_true('loopback byte matches', data[0] == 0x42)

result = transport.write_read(bytes([0xA5, 0x5A]), 2)
check_true('write_read returns 2 bytes', len(result) == 2)
check_true('write_read loopback matches', result == bytes([0xA5, 0x5A]))

print('===DONE: {} passed, {} failed==='.format(passed, failed))
